src/gui/sdlmain.cpp

```diff
@@ -55,6 +56,10 @@
 #include "control.h"
 #include "render.h"
 
+extern "C" {
+#include "../libs/ppscale/ppscale.h"
+}
+
```

Comment:
*dreamer*:

> `extern` should be placed in a header of your library, not in code using your library; see `src/libs/decoders/SDL_sound.h` for example how to handle it (that's the normal way of making your header compatible with both C and C++).

That is how I did it the first time around, but then I found I do not have to contaminate my *C* library if I indicate `extern "C"` at the place of inclusion. It is more natural for the caller to adapt to the callee than the other way round: the library provides a single interface and everybody invokes it according to that interface.

The requirement for the library to adapt to different callers is the wrong direction of dependency. Why do prefer it?

If need be, I can create a `ppscale.hpp` with:

    extern "C" {
    #include "ppscale.h"
    }

Reply:
Just a note on `ppscale.h`. I see that the `asp` parameter to `pp_getscale` is terribly misnamed. I will give it a more sensible name, such as `parweight`—the weight of *PAR* in the estimation of scaling factors.

Reply:
Placing ifdefed `extern "C"` in C library to indicate proper language linking is normal, somewhat expacted from C libraries at this point.

We have several C libraries included and all of them do it (or are ok with being treated as C++). Post-review I will either move this extern block myself or rename your library to .cpp.

Reply:
> Placing ifdefed `extern "C"` in C library to indicate proper language linking is normal, somewhat expacted from C libraries at this point.
> We have several C libraries included and all of them do it (or are ok with being treated as C++). Post-review I will either move this extern block myself or rename your library to .cpp.

I should prefer to keep the *C* header clean. If you dislike the idea of a separate `ppscale.hpp` that wraps the original `ppscale.h` into `extern C`, then I propose to ditch my library altogether, moving `pp_getscale()` directly into `sdlmain.cpp` with a reference to the original library in the comment.

Reply:
Yes, I dislike idea of treating one C library differently than every other C library we use (or in fact, every other C library I ever used in a C++ program over last 20 years or so).

Please note, I am not asking you to change the header of your library any more - if you want to keep your library as C-only, that's fine. I will externs after this PR is merged-in and will deal with merging-in future versions of your lib to dosbox-staging source code; preserving formatting of your lib is going to make it easier.

If you want to drop separate library and just place a new function in dosbox-staging code… that would be ok as well, except it means reformatting your code to K&R (which I presume you don't want), and `sdlmain.cpp` is not good place for it - this file has already too many functions unrelated to main SDL event loop stuffed in there.

Ultimately, it's your call. I think the *fastest* way forward would be to leave it as it is and let me place the extern block.

Reply:
@dreamer:

> Yes, I dislike idea of treating one C library differently than every other C library we use (or in fact, every other C library I ever used in a C++ program over last 20 years or so).

I can volunteer and fix the usage of *all* C libraries in *DOSBox* according to what I see as the proper way: by indicating the *C* calling convention in the calling module either directly or by an intermediate `.hpp` file. Indicating it in *C* files is a sort of inside-out inverted dependency, when a *C* library should be aware, as it were, of *C++*. In my opinion the caller should adapt to the callee, not the other way round.

> Please note, I am not asking you to change the header of your library any more - if you want to keep your library as C-only, that's fine. I will externs after this PR is merged-in and will deal with merging-in future versions of your lib to dosbox-staging source code; preserving formatting of your lib is going to make it easier.

Thank you. One advantage of using a library in unmodified form is ease of maintenance: one can simply copy the file from the repository into the *DOSBox* source. That is why I didn't like your proposal of maintaining a modified copy.

> If you want to drop separate library and just place a new function in dosbox-staging code… that would be ok as well, except it means reformatting your code to K&R (which I presume you don't want), and `sdlmain.cpp` is not good place for it - this file has already too many functions unrelated to main SDL event loop stuffed in there.

I *might* reformat that single function in K&R, but again—it would impose additional maintenance.

> Ultimately, it's your call. I think the _fastest_ way forward would be to leave it as it is and let me place the extern block.

If you insist on putting `exterc "C"` into the `.h` files, then I give up and let you do it after the merge.